spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.reflect.ClassTag
+
+/**
+ * A layer which takes a table of multiple tensors(n >= 2) as input
+ * and calculate to dot product for `all combinations of pairs` among input tensors.
+ * <br><br>
+ * Dot-product outputs are ordered according to orders of pairs in input Table.
+ * For instance, input (Table) is T(A, B, C), output (Tensor) will be [A.*B, A.*C, B.*C].
+ * <br><br>
+ * Dimensions of input' Tensors could be one or two, if two, first dimension is `batchSize`.
+ * For convenience, output is 2-dim Tensor regardless of input' dims.
+ * <br><br>
+ * Table size checking and Tensor size checking will be execute before each forward,
+ * when [[numTensor]] and [[embeddingSize]] are set values greater than zero.
+ *
+ * @param propagateBack whether updateGradInput during backward, default: true
+ * @param numTensor (for checking)number of Tensor input Table contains, default: None
+ * @param embeddingSize (for checking)vector length of dot product, default: None
+ */
+class CrossProduct[T: ClassTag](
+  val propagateBack: Boolean = true,
+  val numTensor: Int = 0,
+  val embeddingSize: Int = 0
+)(implicit ev: TensorNumeric[T]) extends AbstractModule[Table, Tensor[T], T] {
+
+  override def updateOutput(input: Table): Tensor[T] = {
+    val len = input.length()
+    require(numTensor <= 0 || numTensor == len,
+      s"len of input Table($len) not equal to numTensor($numTensor)!")
+
+    val (_, batch, _) = getShape(input[Tensor[T]](1))
+    output.resize(batch, len * (len - 1) / 2)
+
+    if (embeddingSize > 0) {
+      (1 to len).foreach(i => checkEmbeddingSize(input(i)))
+    }
+
+    var cc = 1
+    for( i <- 1 until len; j <- i + 1 to len ) {
```

Comment:
maybe use while is better than for here, as it's in module forward path(compute sensitive). As far as I know there're some overhead in ```for``` in scala.